File: leetcode/leetcode3336.cpp
#include<vector>
#include<algorithm>
#include<numeric>
#include<unordered_map>
using namespace std;

class Solution {
public:
    int subsequencePairCount(vector<int>& nums) {
        const int MOD = 1'000'000'007;
        int n=nums.size();
        int m = *max_element(nums.begin(), nums.end());
        vector<vector<int>> dp(m+1,vector<int>(m+1));
        dp[0][0]=1;
        for(int num:nums){
            vector<vector<int>> ndp(m+1,vector<int>(m+1));
            for(int j=0;j<=m;j++){
                int divisor1 = gcd(j, num);
                for(int k=0;k<=m;k++){
                    int divisor2 = gcd(k, num);
                    ndp[j][k]=(ndp[j][k]+dp[j][k])%MOD;
                    ndp[divisor1][k] = (ndp[divisor1][k] + dp[j][k]) % MOD;
                    ndp[j][divisor2] = (ndp[j][divisor2] + dp[j][k]) % MOD;
                }
            }
            dp=ndp;
        }
        int ans=0;
        for(int i=0;i<=m;i++){
            ans=(ans+dp[i][i])%MOD;
        }
        return ans;
    }
};

File: leetcode/leetcode3336.py
#https://leetcode.cn/problems/find-the-number-of-subsequences-with-equal-gcd/description/?envType=daily-question&envId=2026-07-14
from typing import List
from math import gcd
from collections import defaultdict


class Solution:
    def subsequencePairCount(self, nums: List[int]) -> int:
        MOD = 10 ** 9 + 7
        dp={(0,0):1}
        for num in nums:
            next_dp=defaultdict(int)
            for(g1,g2),count in dp.items():
                next_dp[g1,g2]=(
                    next_dp[g1,g2]+count
                )%MOD
                new_g1=gcd(g1,num)
                next_dp[new_g1,g2]=(
                    next_dp[new_g1,g2]+count
                )%MOD
                new_g2=gcd(g2,num)
                next_dp[g1,new_g2]=(
                    next_dp[g1,new_g2]+count
                )%MOD
            dp=next_dp
        return sum(
            count for(g1,g2),count in dp.items()
            if(g1==g2 and g1>0)
        )%MOD


